substrate/client/network/src/litep2p/shim/request_response/mod.rs

```diff
@@ -337,10 +337,7 @@ impl RequestResponseProtocol {
 					response.len(),
 				);
 
-				let _ = tx.send(Ok((
-					response,
-					fallback.map_or_else(|| self.protocol.clone(), Into::into),
-				)));
+				let _ = tx.send(Ok((response, self.protocol.clone())));
```

Comment:
Unfortunately, this is not a fix for our issue because the fallback mentioned here is actually a _fallback protocol_, and not a _fallback name_. Here we need to pass the actual name of the _fallback protocol_ for the client code to correctly implement protocol backward-compatibility (e.g., correctly process a different protobuf schema).

I have created an issue in litep2p for adding support to actual _fallback names_:
https://github.com/paritytech/litep2p/issues/289

Reply:
To sum up the difference in operation of libp2p & litep2p network backends:

### libp2p network backend
1. libp2p backend is responsible for [sending the fallback request](https://github.com/paritytech/polkadot-sdk/blob/c422d8bbae8ba327597582203f05d30c26ef1392/substrate/client/network/src/request_responses.rs#L977-L994), not libp2p.
2. If only the fallback protocol _name_ was used on the wire and no fallback _request_ happened, the [main protocol name is returned](https://github.com/paritytech/polkadot-sdk/blob/c422d8bbae8ba327597582203f05d30c26ef1392/substrate/client/network/src/request_responses.rs#L845-L850) to the client code
3. If the fallback request was sent, the name of the fallback request protocol is returned to the client code (as it is just another protocol from the libp2p backend perspective).

### litep2p network backend
1. litep2p [sends the fallback request internally](https://github.com/paritytech/litep2p/blob/07240f22cc63baac46e70e3b4ee10f62084b707d/src/protocol/request_response/mod.rs#L359-L364) and it is the one responsible for actually performing the fallback request, not the network backend.

   Surprisingly enough, litep2p network backend also has the logic of [handling fallback requests](https://github.com/paritytech/polkadot-sdk/blob/bb5653eb04fbce40b2b7eb1f1ad62506ea2c4b0a/substrate/client/network/src/litep2p/shim/request_response/mod.rs#L421-L448) in case litep2p returned `UnsupportedProtocol` for some reason.
2. If only the fallback protocol _name_ was used on the wire and no fallback request happened, the fallback protocol name is returned to the client code. This is what makes litep2p different and leads to panics in https://github.com/paritytech/polkadot-sdk/issues/6581.
3. If the fallback request was sent, the name of the fallback request protocol is returned to the client code (the same as libp2p).

I will create a PR to make litep2p/backend behave in a way libp2p backend behaves.

Reply:
> itep2p [sends the fallback request internally](https://github.com/paritytech/litep2p/blob/07240f22cc63baac46e70e3b4ee10f62084b707d/src/protocol/request_response/mod.rs#L359-L364) and it is the one responsible for actually performing the fallback request, not the network backend.

This turned out to be not the case: litep2p network backend does not use the litep2p's [`try_send_request_with_fallback`](https://github.com/paritytech/litep2p/blob/d6fae559fbe2aaf8cb3fff021fb5608ca7b43766/src/protocol/request_response/handle.rs#L459) and uses [`try_send_request`](https://github.com/paritytech/polkadot-sdk/blob/f520adb0aacd51247ed548c7db9bef874c2cca9e/substrate/client/network/src/litep2p/shim/request_response/mod.rs#L248) instead. This means fallback _requests_ are handled completely by the backend in substrate, and not by litep2p.

So, we can be sure that a fallback reported by litep2p is just the fallback _name_ of the main protocol and safely override it with the main protocol name in substrate.

Reply:
Oki doki, thanks a lot for the detailed info and for confirming, this helps a lot🙏